crates/sui-indexer/src/store/pg_indexer_store_v2.rs

```diff
@@ -563,76 +569,98 @@ impl PgIndexerStoreV2 {
-                                epochs::epoch_total_transactions
-                                    .eq(excluded(epochs::epoch_total_transactions)),
-                                epochs::last_checkpoint_id.eq(excluded(epochs::last_checkpoint_id)),
-                                epochs::epoch_end_timestamp
-                                    .eq(excluded(epochs::epoch_end_timestamp)),
-                                epochs::storage_fund_reinvestment
-                                    .eq(excluded(epochs::storage_fund_reinvestment)),
-                                epochs::storage_charge.eq(excluded(epochs::storage_charge)),
-                                epochs::storage_rebate.eq(excluded(epochs::storage_rebate)),
-                                epochs::storage_fund_balance
-                                    .eq(excluded(epochs::storage_fund_balance)),
-                                epochs::stake_subsidy_amount
-                                    .eq(excluded(epochs::stake_subsidy_amount)),
-                                epochs::total_gas_fees.eq(excluded(epochs::total_gas_fees)),
-                                epochs::total_stake_rewards_distributed
-                                    .eq(excluded(epochs::total_stake_rewards_distributed)),
-                                epochs::leftover_storage_fund_inflow
-                                    .eq(excluded(epochs::leftover_storage_fund_inflow)),
-                                epochs::new_total_stake.eq(excluded(epochs::new_total_stake)),
-                                epochs::epoch_commitments.eq(excluded(epochs::epoch_commitments)),
-                                epochs::next_epoch_reference_gas_price
-                                    .eq(excluded(epochs::next_epoch_reference_gas_price)),
-                                epochs::next_epoch_protocol_version
-                                    .eq(excluded(epochs::next_epoch_protocol_version)),
-                            ))
-                            .execute(conn)?;
-                    }
-                    let epoch_id = epoch_data.new_epoch.epoch;
-                    info!(epoch_id, "Persisting initial epoch state");
-                    let new_epoch =
-                        StoredEpochInfo::from_epoch_beginning_info(&epoch_data.new_epoch);
+                if let Some(last_epoch) = &epoch.last_epoch {
+                    let last_epoch_id = last_epoch.epoch;
+                    let last_epoch = StoredEpochInfo::from_epoch_end_info(last_epoch);
+                    info!(last_epoch_id, "Persisting epoch end data: {:?}", last_epoch);
                     diesel::insert_into(epochs::table)
-                        .values(new_epoch)
-                        .on_conflict_do_nothing()
+                        .values(last_epoch)
+                        .on_conflict(epochs::epoch)
+                        .do_update()
```

Comment:
IIUC this updates the row with the new value if a row with this epoch id already exists. Is that right? Does this actually happen in production?

Reply:
Ah I see we are updating both last epoch and the new epoch in this function. So last epoch indeed exists in the table already. Then for the new epoch insertion, why do we `do_nothing`? The new epoch shouldn't have a row in the db yet, right?

Reply:
yeah the intents here are:
- the prev epoch data only have several columns like epoch, starting cp & timestamp populated, all others are default before updating;
- for the new epoch, in theory we do not need do_nothing() afaict, but idempotency is good in case DB gives wrong signals and indexer retries but the row has been there already